File: apps/api/services/password_policy.py
"""One password policy, used by every path that sets a password (§200).

Before this there was no policy at all on the server: `/auth/set-password`,
`/auth/accept-invite`, `/auth/register` and the fresh-install setup route each
took whatever string arrived and hashed it. The only rule anywhere was a
`password.length < 8` check in the browser, in two separate components — which
is not a rule, it is a suggestion to anyone not using that browser.

So this module is deliberately the ONLY place the rules live, and every one of
those call sites now goes through `validate_password`. A second copy of "what
counts as an acceptable password" is the shape this codebase keeps getting
bitten by (§190 found one byte-formatting rule written four times, each wrong
the same way), and here the cost of drift is not a mislabelled megabyte.

The rules, in the order they are checked — the order IS the design, because the
first failure is the message the user reads, and it should be the most concrete
thing that is wrong:

  1. not in the committed list of the most common passwords in the world
  2. at least 12 characters
  3. at least one upper case, one lower case, one digit, one special character
  4. does not contain the local part of either address, the person's name, or
     this instance's name
  5. a zxcvbn strength score of at least MIN_STRENGTH_SCORE

Rule 1 is first because behind the length and class rules it is unreachable —
see `validate_password` for the count that established that.

**Rules 3 and 5 are both required on purpose, and they disagree by design.**

A four-word passphrase of 24 characters — the kind of thing a strength meter
loves, and rightly — is REJECTED here if it has no special character, because
rule 3 applies at every length with no exemption. That is Mathias' explicit
decision (2026-09-20), made knowing it costs some genuinely strong passwords.
It is not an oversight to be tidied up later: anyone reading this and thinking
"a long passphrase should skip the class rule" is re-opening a decision that
was already made, and should ask before changing it.

The score exists for the opposite failure, the one the class rule cannot see:
`Sommer2026!` satisfies all four classes, is in no blocklist, contains nobody's
name, and is a word plus the current year plus the punctuation mark everyone
reaches for. It scores 2 of 4 here (see LOCALE_WORDS for why it takes a
deliberate nudge to score it the way `Summer2026!` already scored). Four
classes are a floor, not evidence.

**The boundary is real and worth knowing.** `Sommer2026!!` — the same password
padded to twelve characters with a second `!` — reaches score 3 and IS
accepted. Score 3 is zxcvbn's own "safely unguessable" (10^8-10^10 guesses)
and is what this policy takes; raising the bar to 4 would refuse a great many
passwords people would actually keep, and the twelve-character minimum plus
the four classes is the floor underneath it.
"""

from __future__ import annotations

import re
from functools import lru_cache
from pathlib import Path
from typing import Optional

from zxcvbn import zxcvbn

#: Twelve, not eight. Eight was the browser-side number and it was never
#: enforced anywhere that mattered.
MIN_PASSWORD_LENGTH = 12

#: zxcvbn scores 0-4. 3 is "safely unguessable" in its own terms — 10^8 to
#: 10^10 guesses — and is the level at which `Sommer2026!` (score 1) and
#: `Summer2026!` (score 1) fail while something genuinely unpredictable of
#: the same length passes.
MIN_STRENGTH_SCORE = 3

#: The frequency-ordered blocklist, committed as a data file. See its own
#: header for provenance. A path constant rather than an inline read so a
#: test can point at a small fixture without monkeypatching the loader.
COMMON_PASSWORDS_PATH = Path(__file__).resolve().parent.parent / "data" / "common_passwords.txt"

#: Words zxcvbn does not know, on an instance whose users do not think in
#: English (§200).
#:
#: This is a measured fix, not a guess. zxcvbn ships English frequency
#: dictionaries, so it scores `Summer2026!` at 2 — correctly recognising
#: word-plus-year — and `Sommer2026!` at 3, because the German word is not in
#: any list it holds. Same password, same pattern, same user; one point apart
#: purely because of which language they happen to think in. Feeding these
#: through `user_inputs` puts them in front of the same matcher that already
#: catches the English half, and `Sommer2026!` drops to 2 where it belongs.
#:
#: Deliberately SHORT, and deliberately not a German word list. A large
#: dictionary here would slow every keystroke on the client's twin and would
#: start refusing ordinary passphrase words; what is wanted is the handful of
#: words a person actually reaches for when a form demands a password —
#: seasons, months, and the words that mean "password". The instance's own
#: name is added per-call from site settings, not here.
#:
#: Not a blocklist: these are not forbidden substrings. They lower the SCORE
#: when a password is mostly made of them, which is the honest thing to do
#: about a word that is common rather than secret.
LOCALE_WORDS = (
    "sommer", "winter", "herbst", "fruehling", "frühling",
    "jaenner", "jänner", "januar", "februar", "maerz", "märz", "april",
    "mai", "juni", "juli", "august", "september", "oktober", "november",
    "dezember",
    "passwort", "kennwort", "geheim", "willkommen", "servus", "schatz",
    "liebe", "sonne", "hallo", "danke",
    "wien", "oesterreich", "österreich", "austria",
    "freeframe",
)

#: Only tokens of at least this length are treated as "contained in the
#: password" for rule 4. Without a floor, a user whose last name is "Li" or
#: whose instance is called "AB" could not use a password containing those
#: two letters in sequence anywhere — which rejects an enormous number of
#: perfectly good passwords for no security gain.
MIN_CONTAINED_TOKEN_LENGTH = 3


class PasswordPolicyError(ValueError):
    """A password was refused, with the reason to show the person.

    Carries a plain sentence rather than a code: there is exactly one
    consumer (an HTTP 400 whose detail the browser renders verbatim), and a
    code would mean the same sentences written again in TypeScript, where
    they would drift from these.
    """

    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


@lru_cache(maxsize=1)
def _common_passwords() -> frozenset[str]:
    """The blocklist, lowercased, read once per process.

    Cached because this is on the path of every password set, and a 76KB
    file re-read per call is pure waste. `lru_cache` rather than a module
    global so a test can call `.cache_clear()` after pointing the path
    somewhere else.

    A missing file is NOT silently tolerated. The alternative — return an
    empty set and carry on — is a policy that quietly stops enforcing one of
    its five rules because of a packaging mistake, and nothing would ever
    report it. The Dockerfile copies `apps/api` wholesale, so the file
    travels with the code; if it ever does not, this fails loudly at the
    first password set rather than at the first breach.
    """
    with COMMON_PASSWORDS_PATH.open(encoding="utf-8") as fh:
        return frozenset(
            line.strip().lower()
            for line in fh
            if line.strip() and not line.startswith("#")
        )


def _local_part(address: Optional[str]) -> Optional[str]:
    if not address or "@" not in address:
        return None
    local = address.split("@", 1)[0].strip().lower()
    return local or None


def _name_tokens(name: Optional[str]) -> list[str]:
    """Word-ish pieces of a name, long enough to be worth checking.

    Split on anything non-alphanumeric so "Mary-Jane" and "O'Brien" yield
    their parts, which is what somebody building a password out of their own
    name would actually use.
    """
    if not name:
        return []
    return [
        part.lower()
        for part in re.split(r"[^0-9A-Za-z]+", name)
        if len(part) >= MIN_CONTAINED_TOKEN_LENGTH
    ]


def personal_tokens(
    *,
    email: Optional[str] = None,
    backup_email: Optional[str] = None,
    name: Optional[str] = None,
    org_name: Optional[str] = None,
) -> list[str]:
    """Everything about this person the password must not contain.

    Shared with `strength_hints` below so the advisory score is computed
    against the same inputs the rejection is, rather than a shorter list
    that would make the meter more optimistic than the server.
    """
    tokens: list[str] = []
    for address in (email, backup_email):
        local = _local_part(address)
        if local and len(local) >= MIN_CONTAINED_TOKEN_LENGTH:
            tokens.append(local)
    tokens.extend(_name_tokens(name))
    tokens.extend(_name_tokens(org_name))
    # Deduplicated, order preserved: the first match is the one reported, and
    # reporting "your email address" twice helps nobody.
    seen: set[str] = set()
    unique: list[str] = []
    for token in tokens:
        if token not in seen:
            seen.add(token)
            unique.append(token)
    return unique


def _strength_reason(result: dict) -> str:
    """zxcvbn's own explanation, turned into one sentence.

    Its `warning` is the concrete finding ("This is a top-100 common
    password", "Names and surnames by themselves are easy to guess") and is
    what makes the rejection actionable; the first suggestion is the fallback
    when there is no warning, which happens for passwords that are merely
    short on entropy rather than recognisably patterned.
    """
    feedback = result.get("feedback") or {}
    warning = (feedback.get("warning") or "").strip()
    if warning:
        return warning.rstrip(".")
    suggestions = feedback.get("suggestions") or []
    if suggestions:
        return str(suggestions[0]).rstrip(".")
    return "It is too easy to guess"


def strength_hints(
    password: str,
    *,
    email: Optional[str] = None,
    backup_email: Optional[str] = None,
    name: Optional[str] = None,
    org_name: Optional[str] = None,
) -> dict:
    """Score and reason, without deciding anything.

    Exists so a caller can report the score alongside a rejection, and so the
    score is computed in exactly one way. The browser runs its own copy of
    zxcvbn for the live meter (`lib/password-policy.ts`) — a different
    implementation of the same algorithm, so the two can differ by a point at
    the boundary. That is tolerable precisely because the meter is advisory
    and this module decides.
    """
    result = zxcvbn(
        # zxcvbn is O(n^2)-ish in the length of the input and a pasted blob
        # should not be able to stall a worker. 128 is far past anything the
        # rules below can distinguish anyway.
        password[:128],
        user_inputs=list(LOCALE_WORDS)
        + personal_tokens(
            email=email, backup_email=backup_email, name=name, org_name=org_name
        ),
    )
    return {"score": int(result.get("score", 0)), "reason": _strength_reason(result)}


def validate_password(
    password: str,
    *,
    email: Optional[str] = None,
    backup_email: Optional[str] = None,
    name: Optional[str] = None,
    org_name: Optional[str] = None,
) -> None:
    """Accept the password, or raise PasswordPolicyError with the reason.

    Every argument but `password` is optional and every caller should pass
    what it has. They are not decoration: rule 4 is the only one that can
    catch `mathias2026!YonStudio` on Mathias' own instance, and it can only
    do that if it is told who is typing and where.

    No rotation, no expiry, deliberately — both push people towards
    `Sommer2026!` and then `Herbst2026!`, which is the exact failure rule 5
    exists to catch.
    """
    # The blocklist goes FIRST, ahead of length and the character classes,
    # and that ordering was corrected after measuring rather than assumed.
    #
    # Behind them it is dead code: exactly 10 of the 10,001 entries are twelve
    # characters or longer and not one of those has all four character
    # classes, so a password that reached this check had already been refused
    # by an earlier rule every single time. A rule that can never fire is not
    # a rule.
    #
    # It is also the better message. Somebody who typed `password` is told
    # what is actually wrong with it — it is one of the most guessed strings
    # in the world — rather than "use at least 12 characters", which invites
    # `password1234`.
    if password.lower() in _common_passwords():
        raise PasswordPolicyError(
            "This is one of the most commonly used passwords in the world. "
            "Choose something else."
        )

    # Then length: of the remaining rules it is the one a person is most
    # likely to be failing, and "too short" is more useful than "no digit"
    # about a four-character password that is also too short.
    if len(password) < MIN_PASSWORD_LENGTH:
        raise PasswordPolicyError(
            f"Use at least {MIN_PASSWORD_LENGTH} characters."
        )

    missing: list[str] = []
    if not any(c.isupper() for c in password):
        missing.append("an upper-case letter")
    if not any(c.islower() for c in password):
        missing.append("a lower-case letter")
    if not any(c.isdigit() for c in password):
        missing.append("a digit")
    # "Special" is the complement of alphanumeric rather than a fixed
    # punctuation set, so a non-Latin script or a typographic character
    # counts. A fixed set would reject a password that is objectively harder
    # to guess than one containing `!`.
    #
    # WHITESPACE IS EXCLUDED, and that is the whole reason the passphrase rule
    # bites. `not c.isalnum()` alone counts the spaces in "Correct Battery
    # Horse Staple 7" as special characters, so every spaced passphrase would
    # satisfy this rule for free — and the decision this policy exists to
    # implement is precisely that a passphrase does NOT get an exemption.
    # Found by measurement: that passphrase was accepted before this line said
    # `and not c.isspace()`.
    if not any(not c.isalnum() and not c.isspace() for c in password):
        missing.append("a special character")
    if missing:
        # Stated in full, not one at a time: a user fixing four rules one
        # round-trip each is how people end up at `Password1!`.
        raise PasswordPolicyError(
            "Add " + ", ".join(missing) + ". All four are required at every "
            "length, including long passphrases."
        )

    lowered = password.lower()
    for token in personal_tokens(
        email=email, backup_email=backup_email, name=name, org_name=org_name
    ):
        if token in lowered:
            raise PasswordPolicyError(
                f"Your password must not contain “{token}” — anyone "
                "targeting you already knows it."
            )

    hints = strength_hints(
        password,
        email=email,
        backup_email=backup_email,
        name=name,
        org_name=org_name,
    )
    if hints["score"] < MIN_STRENGTH_SCORE:
        raise PasswordPolicyError(
            f"This password is too easy to guess. {hints['reason']}."
        )


def describe_policy() -> dict:
    """The rules as data, for a client that wants to state them up front.

    Kept here rather than written out again in the browser so the screen and
    the server cannot disagree about what is required.
    """
    return {
        "min_length": MIN_PASSWORD_LENGTH,
        "min_strength_score": MIN_STRENGTH_SCORE,
        "requires_upper": True,
        "requires_lower": True,
        "requires_digit": True,
        "requires_special": True,
    }


def validate_password_for_user(password: str, user, org_name: Optional[str] = None) -> None:
    """`validate_password` with the arguments pulled off a User row.

    A convenience with a real purpose: four routers set passwords, and three
    of them would otherwise each assemble this argument list by hand. One that
    forgets `backup_email` is a rule that silently stops applying.
    """
    validate_password(
        password,
        email=getattr(user, "email", None),
        backup_email=getattr(user, "backup_email", None),
        name=getattr(user, "name", None),
        org_name=org_name,
    )

File: apps/web/lib/password-policy.ts
/**
 * The live password meter (§200).
 *
 * **Advisory only. `apps/api/services/password_policy.py` decides.** This
 * module exists so somebody typing a password finds out it is weak while they
 * are typing it, instead of after a round-trip — not so the browser can hold
 * an opinion the server does not share. Every rule here has a twin on the
 * server, the server's twin runs on every path that sets a password, and a
 * password this file happens to like is still refused if the server's does
 * not.
 *
 * The two ARE different implementations — Python `zxcvbn` 4.4 there,
 * `@zxcvbn-ts/core` 3.x here — so the score can differ by a point at the
 * boundary. That is tolerable precisely because of the sentence above. What
 * would not be tolerable is this file inventing a scorer of its own: a
 * hand-rolled "has a digit, has a symbol, therefore strong" meter is how
 * `Sommer2026!` gets a green bar.
 *
 * The four class rules are duplicated here rather than fetched, because they
 * have to run on every keystroke. The NUMBERS are not duplicated —
 * `GET /auth/password-policy` serves the minimum length and score, and
 * `usePasswordPolicy` below reads them from there so the screen and the
 * server cannot disagree about what is required.
 */

import * as React from 'react'
import useSWR from 'swr'
import { api } from '@/lib/api'
import type { PasswordPolicy, PasswordStrength } from '@/types'

/**
 * The fallback used until `GET /auth/password-policy` answers, and if it
 * never does.
 *
 * It matches the server's constants as of §200. Stating it as a fallback
 * rather than as the source of truth is the point: if the two ever drift, the
 * screen is briefly wrong about the minimum length, which costs one rejected
 * submission — whereas treating this as authoritative would let the screen be
 * permanently wrong and confident.
 */
export const DEFAULT_PASSWORD_POLICY: PasswordPolicy = {
  min_length: 12,
  min_strength_score: 3,
  requires_upper: true,
  requires_lower: true,
  requires_digit: true,
  requires_special: true,
}

/**
 * Read a dynamically imported module's real exports, whichever shape the
 * bundler hands back (§202).
 *
 * This function exists because of a bug that reached production and was
 * invisible to the whole test suite. The original code read
 * `common.default.dictionary`, and `@zxcvbn-ts/language-*` has **no default
 * export at all** — its ESM build ends in `export { adjacencyGraphs,
 * dictionary }`. In a true ESM namespace `.default` is therefore `undefined`,
 * and `.default.dictionary` throws a TypeError.
 *
 * It passed every test anyway. Vitest resolves these packages' CJS build and
 * synthesises `default = module.exports`, which happens to carry `.dictionary`
 * — so `common.default.dictionary` works in jsdom and only ever fails in the
 * browser, where Next resolves the `module` field and gets the real namespace.
 * A test environment that is kinder than production is worse than no test.
 *
 * So this picks by CONTENT rather than by shape: whichever of the namespace or
 * its `.default` actually carries the keys being asked for. That is stable
 * across both interops, and across the package growing a default export later.
 */
export function resolveModuleExports<T>(ns: unknown, ...expectedKeys: string[]): T {
  const candidates = [ns, (ns as { default?: unknown })?.default]
  for (const candidate of candidates) {
    if (
      candidate &&
      typeof candidate === 'object' &&
      expectedKeys.every((key) => key in (candidate as object))
    ) {
      return candidate as T
    }
  }
  // Nothing carried them. Throwing names the module shape in the message
  // instead of letting a TypeError surface three frames away as
  // "cannot read properties of undefined".
  throw new Error(
    `zxcvbn module is missing ${expectedKeys.join(', ')}; got keys: ` +
      `${ns && typeof ns === 'object' ? Object.keys(ns).join(', ') : typeof ns}`,
  )
}

interface ZxcvbnLanguage {
  dictionary: Record<string, (string | number)[]>
  adjacencyGraphs?: Record<string, unknown>
  translations?: unknown
}

/**
 * zxcvbn's dictionaries, loaded once and only when a password field is
 * actually rendered.
 *
 * `@zxcvbn-ts/language-common` is ~800KB of frequency lists. Importing it at
 * module scope would put that in the bundle of every page that transitively
 * reaches this file — including the dashboard, which never asks for a
 * password. A dynamic import inside a memoised promise keeps it on the two
 * screens that need it and loads it exactly once per session.
 */
let optionsPromise: Promise<void> | null = null

/**
 * Turn the two imported language modules into the options object zxcvbn wants.
 *
 * Exported and pure ONLY so it can be tested, and that is not a small point:
 * the §202 bug lived in exactly these four lines, in code that could not be
 * reached from a test because it sat inside a dynamic import. A test can now
 * hand this the real ESM namespace — the one with no default export, the one
 * the browser gets — and assert it produces a populated dictionary. Reverting
 * to `commonNs.default.dictionary` fails that test instead of shipping.
 */
export function buildZxcvbnOptions(commonNs: unknown, enNs: unknown) {
  const common = resolveModuleExports<ZxcvbnLanguage>(
    commonNs,
    'dictionary',
    'adjacencyGraphs',
  )
  const en = resolveModuleExports<ZxcvbnLanguage>(enNs, 'dictionary', 'translations')
  return {
    dictionary: { ...common.dictionary, ...en.dictionary },
    graphs: common.adjacencyGraphs as never,
    translations: en.translations as never,
  }
}

async function ensureZxcvbnOptions(): Promise<
  typeof import('@zxcvbn-ts/core')
> {
  const core = await import('@zxcvbn-ts/core')
  if (!optionsPromise) {
    optionsPromise = (async () => {
      const [commonNs, enNs] = await Promise.all([
        import('@zxcvbn-ts/language-common'),
        import('@zxcvbn-ts/language-en'),
      ])
      core.zxcvbnOptions.setOptions(buildZxcvbnOptions(commonNs, enNs))
    })()
    // §202 — a REJECTED promise must not be cached.
    //
    // Left as-is, one failure (a dropped chunk on a flaky connection, a
    // transient 502 from the CDN) poisons the meter for the rest of the
    // session: every later call awaits the same rejected promise and returns
    // instantly without retrying. That is how a momentary network blip
    // becomes a permanently dead password form.
    optionsPromise.catch(() => {
      optionsPromise = null
    })
  }
  await optionsPromise
  return core
}

/**
 * The same short locale list `services/password_policy.py` feeds to its own
 * zxcvbn, kept in step by hand.
 *
 * zxcvbn's dictionaries are English, so `Summer2026!` scores 2 and
 * `Sommer2026!` scores 3 — one point apart purely because of which language
 * the user thinks in. Without this the meter would be a full point more
 * optimistic than the server for exactly the passwords the people on this
 * instance actually choose, which is the one direction a meter must never be
 * wrong in.
 *
 * Duplicated rather than fetched because it runs on every keystroke and the
 * list is short and stable. If the server's list grows, grow this one: a
 * drift here shows up as a meter that reads "medium" on something the server
 * refuses, which is annoying rather than dangerous — but it is still wrong.
 */
const LOCALE_WORDS = [
  'sommer', 'winter', 'herbst', 'fruehling', 'frühling',
  'jaenner', 'jänner', 'januar', 'februar', 'maerz', 'märz', 'april',
  'mai', 'juni', 'juli', 'august', 'september', 'oktober', 'november',
  'dezember',
  'passwort', 'kennwort', 'geheim', 'willkommen', 'servus', 'schatz',
  'liebe', 'sonne', 'hallo', 'danke',
  'wien', 'oesterreich', 'österreich', 'austria',
  'freeframe',
]

/** Which of the four character classes are still missing. */
export function missingClasses(
  password: string,
  policy: PasswordPolicy,
): string[] {
  const missing: string[] = []
  if (policy.requires_upper && !/[A-Z]/.test(password)) missing.push('an upper-case letter')
  if (policy.requires_lower && !/[a-z]/.test(password)) missing.push('a lower-case letter')
  if (policy.requires_digit && !/[0-9]/.test(password)) missing.push('a digit')
  // The complement of alphanumeric MINUS whitespace, matching the server's
  // own definition exactly. Both halves matter: not a fixed punctuation set,
  // because that would show a red mark for a password the server accepts; and
  // not counting spaces, because `[^A-Za-z0-9]` alone treats the spaces in a
  // passphrase as special characters — which would let this meter say a
  // spaced passphrase is fine right up until the server refuses it.
  if (policy.requires_special && !/[^A-Za-z0-9\s]/.test(password)) {
    missing.push('a special character')
  }
  return missing
}

export const STRENGTH_LABELS = ['weak', 'weak', 'weak', 'medium', 'strong'] as const

/**
 * Score a password, the same way and against the same inputs the server does.
 *
 * `userInputs` is not optional in spirit: the server rejects a password
 * containing the person's own address or name outright, and zxcvbn scores
 * those far lower when it is told about them. Calling this without them
 * produces a meter that is more optimistic than the server, which is the one
 * direction a meter must never be wrong in.
 */
export async function scorePassword(
  password: string,
  userInputs: string[],
  policy: PasswordPolicy = DEFAULT_PASSWORD_POLICY,
): Promise<PasswordStrength> {
  if (!password) {
    return { score: 0, label: 'weak', reason: '', meetsPolicy: false, missing: [] }
  }

  const missing = missingClasses(password, policy)
  const tooShort = password.length < policy.min_length

  const core = await ensureZxcvbnOptions()
  // Capped at 128 like the server, and for the same reason: zxcvbn's cost
  // grows fast with length, and this runs on every keystroke.
  const result = core.zxcvbn(password.slice(0, 128), [
    ...LOCALE_WORDS,
    ...userInputs,
  ])
  const score = result.score as number

  // zxcvbn's own finding, which is the part that makes a weak score
  // actionable — "This is similar to a commonly used password" tells the
  // person what to change; a bar that is 40% full does not.
  const reason =
    result.feedback.warning?.replace(/\.$/, '') ??
    result.feedback.suggestions[0]?.replace(/\.$/, '') ??
    ''

  return {
    score,
    label: STRENGTH_LABELS[Math.min(Math.max(score, 0), 4)],
    reason,
    // Everything the browser can check. Deliberately NOT called "valid": the
    // blocklist and the personal-token rule only exist server-side, so a
    // password that passes here can still be refused, and a name suggesting
    // otherwise would invite a caller to skip handling that.
    meetsPolicy:
      !tooShort && missing.length === 0 && score >= policy.min_strength_score,
    missing,
  }
}

/**
 * Report a scoring failure once per session, loudly enough to be found.
 *
 * Once, because this runs on a debounce behind every keystroke and a
 * per-attempt log would bury the first and only useful one under a hundred
 * copies. Loudly, because the alternative is what §202 was: a `catch` that
 * returned null, a submit button that never enabled, and no evidence at all
 * in the console, the network tab or the server logs.
 */
let strengthFailureReported = false

function reportStrengthFailure(err: unknown): void {
  if (strengthFailureReported) return
  strengthFailureReported = true
  // eslint-disable-next-line no-console
  console.error(
    '[password-policy] strength scoring is unavailable; the meter will stay ' +
      'blank and the server will enforce the policy on submit.',
    err,
  )
}

/**
 * Why the submit button cannot be pressed yet — or `null` when it can (§202).
 *
 * ONE rule, shared by the four forms that set a password (invite acceptance,
 * the §200 onboarding gate, the login screen's set-password step, and
 * Settings → Profile). They had four copies of
 * `disabled={!strength?.meetsPolicy || !confirm}`, which is how all four
 * inherited the same silent failure at once.
 *
 * Two principles, and they are the actual fix rather than the interop patch:
 *
 * **A blocked submit always has a sentence.** The caller renders the returned
 * string next to the button. There is no state in which the control is dead
 * and the screen says nothing — that is the bug class §202 is about, not the
 * particular TypeError that caused it this time.
 *
 * **An unknown score never blocks.** `strength` is null while the debounce is
 * pending, while the 800KB dictionary is still downloading, and forever if
 * that download fails. None of those are statements about the password. The
 * server runs the real policy on every path that sets one — including the
 * blocklist and the personal-token rule, which the browser cannot check at
 * all — so letting an unscored password through costs a round trip and a
 * rendered error, while blocking it costs the user their account.
 *
 * What DOES block is only what the browser can be certain of: the two fields
 * disagreeing, and the length and character-class rules, which are computed
 * synchronously and need no dictionary at all.
 */
export function passwordSubmitBlock(args: {
  password: string
  confirmPassword: string
  strength: PasswordStrength | null
  policy: PasswordPolicy
}): string | null {
  const { password, confirmPassword, strength, policy } = args

  if (!password) return 'Enter a password.'

  // Synchronous, dictionary-free, and therefore always available — this is
  // the part that can safely block.
  if (password.length < policy.min_length) {
    return `Use at least ${policy.min_length} characters.`
  }
  const missing = missingClasses(password, policy)
  if (missing.length > 0) {
    return `Add ${missing.join(', ')}.`
  }

  if (!confirmPassword) return 'Confirm your password.'
  if (password !== confirmPassword) return 'Passwords do not match.'

  // Null means "not scored yet", never "bad". See the note above.
  if (strength && strength.score < policy.min_strength_score) {
    return strength.reason
      ? `This password is too easy to guess. ${strength.reason}.`
      : 'This password is too easy to guess.'
  }

  return null
}

/** The instance's rules, fetched once. Falls back to the constants above. */
export function usePasswordPolicy(): PasswordPolicy {
  const { data } = useSWR<PasswordPolicy>(
    '/auth/password-policy',
    () => api.get<PasswordPolicy>('/auth/password-policy'),
    {
      // The rules do not change between renders, and a screen that refetched
      // them on every focus would be spending requests to be told the same
      // thing.
      revalidateOnFocus: false,
      revalidateIfStale: false,
    },
  )
  return data ?? DEFAULT_PASSWORD_POLICY
}

/**
 * Debounced scoring for a controlled input.
 *
 * Debounced because `scorePassword` awaits a dynamic import on its first call
 * and then runs a real analysis; doing that synchronously per keystroke makes
 * a password field feel heavy. 150ms is below the threshold where the meter
 * reads as lagging, and above the one where it re-runs for every character of
 * a fast typist.
 */
export function usePasswordStrength(
  password: string,
  userInputs: string[],
  policy: PasswordPolicy,
): PasswordStrength | null {
  const [strength, setStrength] = React.useState<PasswordStrength | null>(null)
  // Joined into a primitive so the effect does not re-run because the caller
  // built a fresh array with the same contents — which is what an inline
  // `[user.email, user.name]` does on every render.
  const inputsKey = userInputs.join('\u0000')

  React.useEffect(() => {
    if (!password) {
      setStrength(null)
      return
    }
    let cancelled = false
    const timer = setTimeout(() => {
      scorePassword(password, inputsKey ? inputsKey.split('\u0000') : [], policy)
        .then((next) => {
          // Guards against an out-of-order resolve overwriting a newer score
          // — the dynamic import makes the first call much slower than the
          // rest, so this is a real ordering, not a theoretical one.
          if (!cancelled) setStrength(next)
        })
        .catch((err) => {
          // A failed dictionary load must not break the form: the meter
          // simply does not appear, and `passwordSubmitBlock` treats a null
          // score as "unknown", which does NOT block submission — the server
          // enforces the real policy either way.
          //
          // §202 — but it is no longer SILENT. This catch is what hid a
          // TypeError for an entire deploy: every password form in the app
          // had a permanently disabled submit button and nothing, anywhere,
          // said why. One console line is the difference between "the button
          // is broken" and a stack trace pointing at the cause.
          reportStrengthFailure(err)
          if (!cancelled) setStrength(null)
        })
    }, 150)
    return () => {
      cancelled = true
      clearTimeout(timer)
    }
  }, [password, inputsKey, policy])

  return strength
}
